_scripts/update.py

```diff
@@ -0,0 +1,179 @@
+from filecmp import cmp
+from os import makedirs
+from os.path import abspath, dirname, exists, join, relpath
+from re import match, sub
+from shutil import copy
+from sys import exit
+from compare_locales.merge import merge_channels
+from compare_locales.parser import Entity, getParser
+from compare_locales.paths import ProjectFiles, TOMLParser
+
+HEAD = "master"
+
+description = f"""
+Update the localization source files from a Firefox branch, adding new files and messages.
+For updates from the "{HEAD}" branch, also update changed messages.
+
+Writes a summary of the branch's localized files and message keys as `_data/[branch].json`,
+and a commit message summary as `.update_msg`.
+"""
+
+
+def add_config(fx_root: str, fx_cfg_path: str, done: set[str]):
+    parts = fx_cfg_path.split("/")
+    if (
+        "{" in fx_cfg_path
+        or len(parts) < 3
+        or parts[-2] != "locales"
+        or parts[-1] != "l10n.toml"
+    ):
+        exit(f"Unsupported config path: {fx_cfg_path}")
+
+    cfg_path = join("_configs", f"{'-'.join(parts[:-2])}.toml")
+    if cfg_path not in done:
+        done.add(cfg_path)
+        with open(join(fx_root, fx_cfg_path), "rb") as file:
+            cfg = tomllib.load(file)
+        cfg["basepath"] = ".."
+        for path in cfg["paths"]:
+            path["reference"] = sub(r"{\s*\S+\s*}", "", path["l10n"])
+        if "includes" in cfg:
+            for incl in cfg["includes"]:
+                incl["path"] = add_config(fx_root, incl["path"], done)
+
+        makedirs(dirname(cfg_path), exist_ok=True)
+        with open(cfg_path, "wb") as file:
+            tomli_w.dump(cfg, file)
+    return cfg_path
+
+
+def update_str(branch: str, fx_root: str, config_files: list[str]):
+    if branch not in [HEAD, "beta", "release"] and not match("esr[0-9]+", branch):
+        exit(f"Unknown branch: {branch}")
+    if not exists(fx_root):
+        exit(f"Firefox root not found: {fx_root}")
+    print(f"source: {branch} at {fx_root}")
+
+    configs = []
+    fixed_configs = set()
+    for cfg_name in config_files:
+        cfg_path = join(fx_root, cfg_name)
+        if not exists(cfg_path):
+            exit(f"Config file not found: {cfg_path}")
+        configs.append(TOMLParser().parse(cfg_path))
+        if branch == HEAD:
+            add_config(fx_root, cfg_name, fixed_configs)
+
+    messages = {}
+    new_files = 0
+    updated_files = 0
+    for fx_path, *_ in ProjectFiles(None, configs):
+        rel_path = relpath(fx_path, abspath(fx_root)).replace("/locales/en-US", "")
+        makedirs(dirname(rel_path), exist_ok=True)
+
+        try:
+            fx_parser = getParser(fx_path)
+        except UserWarning:  # No parser found
+            messages[rel_path] = []
+            if not exists(rel_path):
+                print(f"create {rel_path}")
+                copy(fx_path, rel_path)
+                new_files += 1
+            elif branch == HEAD and not cmp(fx_path, rel_path):
```

Comment:
I'm probably missing some context, but it seems we don't do anything for an updated file if it is unparsable for branches other than HEAD. Is this okay?

I see that we handle that case with parseable files below.

Reply:
This is handling e.g. `browser/firefox-l10n.js`, and the current behaviour is for the changes in central (aka `HEAD` here) to get reflected directly in gecko-strings, ignoring any content on other branches. Hence we do the same here.